### ISSUE TITLE
Stack Level Too Deep - Circular Code Path with around_save + relation update, and has_one

### ISSUE THREAD

**Issue by mordof-work** [rails/rails#28908]
### Steps to reproduce

I created a small environment program that demonstrates the problem. The README.md explains more about how to setup, and run the test environment.

https://github.com/mordof-work/test_environment

### Expected behavior
I expect that when an agency record gets saved with a changed name/is_deleted flag, the associated node record has those fields set, and gets saved. Then everything moves on business as usual.

### Actual behavior
Instead of carrying on after node.save happens - node.save triggers another agency update. The <x>_changes? knowledge never gets resolved, which causes the node to always get saved again and again.

This causes a circular code path, resulting in a Stack Level Too Deep.

### System configuration
**Rails version**:
ActiveRecord / ActiveModel / ActiveSupport version 5.0.2

Problem is not found in ActiveRecord / ActiveModel / ActiveSupport 4.1.5 (which is what I'm upgrading from)

**Ruby version**:
2.3.3

**Comment by mordof-work** [comment]
Referencing my repository - models/agency.rb - lines 26-32 leave a comment about how line 33:  if node.present?   is critical here.

It seems the order in which after_create gets called in callbacks in relation to around_save is changed between ActiveRecord (or ActiveModel, whichever is responsible) 4.1.2, and 5.0.2.

In 5.0.2, update_node gets called first, and the first pass through node.present? is false for the entirety of update_node. It doesn't try and save the related node record. Then the after_create gets called for link_node. This updates node_id in the Agency, which triggers another save. Here's where the issue is.

In 4.1.2, node.present? is false for the second update_node call (which was triggered by create_node).
In 5.0.2, node.present? is true for the second update_node call.

Now that I think about it, this may have been a broken behaviour. node should be available here, since a node was just created.

Can anyone confirm that this was filed as a bug, or something that was intentionally updated for rails after 4.1.2? That the around_save would have knowledge of updated associations that are set in that save itself?

I'll continue digging into this tomorrow to see if I can identify the answer to that in changelogs, etc.

I'm still not sure how update_node wouldn't have caused a circular save in 4.1.2 when an agency got updated with a new name though.

**Comment by mordof-work** [comment]
tl;dr for my post before:

node.present? in update_node (under v4.1.2) was false. I suspect this is a bugfix, because the node was created right before that got hit - node should be present.

<x>_changed? now doesn't get cleared in v5.0.2 like it did before. Did the callback chain get updated so changed tracking happens in a different place?

These are the two things that are different, causing my stack level too deep. If they are intentional updates to Rails 5, great. My situation is a bad use case - I can just use an after_update. If these are introduced bugs, I'd like to help with being able to submit a proper ticket for it.

**Comment by mordof-work** [comment]
I tried using an after_update for the same check to update the node - it still triggers Stack Level Too Deep.

**Comment by mordof-work** [comment]
This is still a problem, but I gave up digging through the code - needed to get some progress on the rails 5 upgrade. 

I put in a workaround, which is in the branch 'functioning_workaround'. It uses an instance variable to keep track of what properties were going to be saved on the node, and if it tries to save the node again with the same details - returns before attempting.

**Comment by mordof-work** [comment]
Confirmed this is working on 5-1-stable. I checked out my test environment project, updated the gems, and ran the code in question: everything worked as expected without stack overflow issues.

This ticket has been resolved.
